Added ability to read in a psd in compressed_bank to be used for calculating overlaps

soumide1102 opened the pull request:
@cdcapano I've added in the ability to read in a psd file in `pycbc_compress_bank` and then use the psd in `compress_waveform` in compress.py for calculating the overlaps. 
I've kept this PR separate from https://github.com/ligo-cbc/pycbc/pull/1209 . Also this PR does not contain the bug fixes I had made while working on the implementation of compressed waveforms in pycbc_inspiral. I'll push those changes in a separate PR after seeing the results of my run.
I tested it for two templates in the O1 bank using the example_psd.txt file available in pycbc for the input psd file and it works fine. The mismatch compared to the case when not using a psd is negligible ~ 0.00001 . I've pasted below the command-line input for reference.
```
pycbc_compress_bank --bank-file H1L1-BANK2HDF-1128299417-1083600.hdf --output test_psd.hdf --low-frequency-cutoff 30 --approximant 'TaylorF2:mtotal<4' 'SEOBNRv2_ROM_DoubleSpin:else' --sample-rate 4096 --tmplt-index 3 5 --compression-algorithm spa --precision double --psd-from-file /home/soumi.de/pycbc-dev-sugwg012417/src/pycbc/examples/psd/example_psd.txt --verbose --force
``` 

Could you please take a look at this?

cdcapano commented on bin/pycbc_compress_bank:
```diff
@@ -136,17 +142,31 @@ for ii in range(imin, imax):
     tmplt = bank.table[ii-imin]
     kmax = htilde.end_idx
-    
+    if args.psd_from_file :
```
Now, down here, you can just use [pycbc.psd.from_cli](https://github.com/ligo-cbc/pycbc/blob/master/pycbc/psd/__init__.py#L28). Note that that function will return None if none of the psd options were set, which is what you want. It will also take care of including a dynamic range factor, which is important when using single precision.

One other thing: it would be faster if you cache the PSDs, instead of regenerating them for every template. You only need a different psd for each `N` that is used. So, create a dictionary to store the psds outside of the loop. Then, when you go to get the PSD:
```
try:
    psd = psd_dict[N]
except KeyError:
   psd = pycbc.psd.from_cli(...)
   psd_dict[N] = psd
```

soumide1102 commented:
@cdcapano Thanks for the comments. I've changed to the `from_cli` option to generate the PSDs. Could you please let me know if the changes look good to you? I've removed the `--sample-rate` argument from `pycbc_compress_bank`, because that would now get inserted via `insert_strain_option_group`. I tested the updated code and have pasted below the command line input below for reference :
```
pycbc_compress_bank --bank-file H1L1-BANK2HDF-1128299417-1083600.hdf --output test_psd.hdf --low-frequency-cutoff 30 --approximant 'TaylorF2:mtotal<4' 'SEOBNRv2_ROM_DoubleSpin:else' --sample-rate 4096 --tmplt-index 3 5 --compression-algorithm spa --precision single --gps-start-time 1126258462 --gps-end-time 1126260462 --frame-type H1_HOFT_C02 --psd-model aLIGOZeroDetHighPower --strain-high-pass 20 --pad-data 8 --channel-name H1:DCS-CALIB_STRAIN_C02 --verbose --force
2017-02-07 15:54:51,365 loading bank
2017-02-07 15:54:51,366 WARNING: no parameters attribute found. Assuming that mass1, mass2, spin1z, spin2z, template_hash are the parameters.
2017-02-07 15:54:51,665 getting needed dfs
2017-02-07 15:54:51,665 writing template info to output
2017-02-07 15:54:51,687 getting compressed amplitude and phase
2017-02-07 15:54:51,694 0: generating TaylorF2 from 30.0 Hz
2017-02-07 15:54:53,159 mismatch: 0.000807, N points: 676 (0 added)
2017-02-07 15:54:53,170 1: generating SEOBNRv2_ROM_DoubleSpin from 30.0 Hz
2017-02-07 15:54:54,434 mismatch: 0.000951, N points: 326 (1 added)
2017-02-07 15:54:54,441 finished
```

cdcapano commented:
@soumide1102 Yeah, I don't think we'll ever use the welch estimate in this code. The reason I said to include the strain option group is that the psd option group includes options about doing a welch estimate from some data, and it doesn't seem right to have some options that just cause an error.

However, I just talked to Alex about it, and we agreed that the better option is to make the welch estimate options optional in `insert_psd_option_group`. So, add a keyword argument to that function, something like `include_data_options` that defaults to True. Only add the `psd-estimation` to `psd-num-segments` options if that keyword argument is True. In `bank_compress` set that option to False. Then you don't need to include the strain option group.

When you make that change to `insert_psd_option_group`, you'll also need to adjust `from_cli` and `verify_psd_options` since `opt.psd_estimation` may not exist.

soumide1102 commented:
@cdcapano Do you want to see something like 
```
if include_data_options :
    psd_options.add_argument("--psd-num-segments", type=int)
```
over here ?
https://github.com/ligo-cbc/pycbc/blob/8248751a5abca80050eb558a653072d342874f19/pycbc/psd/__init__.py#L178
Also, I didn't quite understand what would be required to be changed in `from_cli` and `verify_psd_options` as those functions don't ask for `psd_estimation` to be entered as a required argument. So how would it affect these functions if we don't enter `psd_estimation` as an argument?

cdcapano commented:
@soumide1102 To answer your first question: yes. The second: there's places in `from_cli` and `verify_psd_options` where it checks if `opt.psd_estimation` is set to some value; [here](https://github.com/ligo-cbc/pycbc/blob/8248751a5abca80050eb558a653072d342874f19/pycbc/psd/__init__.py#L86) for example. However, if `psd-estimation` is never added to the parser options by `insert_psd_option_group`, then you'll get an error (I think an `AttributeError`, not sure though) when it tries to do those checks. A way around this in `from_cli` would be to put some code at top, like:
```
try:
    psd_estimation = opt.psd_estimation is not None
except AttributeError:
    psd_estimation = False
```
Then, in the `elif` have `elif psd_estimation` instead of `elif opt.psd_estimation`. You'll need to do similar sorts of things in `verify_psd_options`.

soumide1102 commented:
@cdcapano I made the changes, but I did a git merge accidentally to update with master which messed up the commit history. So I thought it would be better to open a new PR and close this one. I've cherry-picked the appropriate commits from this PR to the new one https://github.com/ligo-cbc/pycbc/pull/1434 .
